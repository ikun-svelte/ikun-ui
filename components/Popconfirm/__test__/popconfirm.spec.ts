import { afterEach, expect, test, describe, beforeEach, vi } from 'vitest';
import KPopconfirmType from './fixture/popconfirm.type.svelte';
import KPopconfirmIcon from './fixture/popconfirm.icon.svelte';
import KPopconfirmDiabled from './fixture/popconfirm.disabled.svelte';
import KPopconfirmCancel from './fixture/popconfirm.cancel.svelte';
import KPopconfirmBtn from './fixture/popconfirm.btn.svelte';
import KPopconfirmBase from './fixture/popconfirm.base.svelte';
import KPopconfirmAsync from './fixture/popconfirm.async.svelte';
import KPopconfirmEvtCancel from './fixture/popconfirm.event.cancel.svelte';
import KPopconfirmEvtChange from './fixture/popconfirm.event.change.svelte';
import KPopconfirmEvtConfirm from './fixture/popconfirm.event.confirm.svelte';
import KPopconfirmSlotFooter from './fixture/popconfirm.slot.footer.svelte';
import KPopconfirmSlotContent from './fixture/popconfirm.slot.content.svelte';
import KPopconfirmSlotTitle from './fixture/popconfirm.slot.title.svelte';
import { tick } from 'svelte';
// TODO E2E test
let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.useRealTimers();
});

describe('Test: KPopconfirm', () => {
	test('props: base', async () => {
		const instance = new KPopconfirmBase({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: icon', async () => {
		const instance = new KPopconfirmIcon({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('i-carbon-settings')).toBeTruthy();
		expect(host.innerHTML.includes('k-popconfirm--error')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: type', async () => {
		const instance = new KPopconfirmType({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elmInfo = host.querySelector('#test_trigger_error');
		elmInfo.click();
		const elmWarning = host.querySelector('#test_trigger_info');
		elmWarning.click();
		const elmSuccess = host.querySelector('#test_trigger_warning');
		elmSuccess.click();
		const elmError = host.querySelector('#test_trigger_success');
		elmError.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('k-popconfirm--error')).toBeTruthy();
		expect(host.innerHTML.includes('k-popconfirm--info')).toBeTruthy();
		expect(host.innerHTML.includes('k-popconfirm--warning')).toBeTruthy();
		expect(host.innerHTML.includes('k-popconfirm--success')).toBeTruthy();
	});

	test('props: disabled', async () => {
		const instance = new KPopconfirmDiabled({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人')).not.toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: showCancel', async () => {
		const instance = new KPopconfirmCancel({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		expect(host.innerHTML.includes('cancel')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: confirmBtnText & confirmBtnCls & cancelBtnText & cancelBtnCls', async () => {
		const instance = new KPopconfirmBtn({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		expect(host.innerHTML.includes('confirmBtnText')).toBeTruthy();
		expect(host.innerHTML.includes('confirmBtnCls')).toBeTruthy();
		expect(host.innerHTML.includes('cancelBtnText')).toBeTruthy();
		expect(host.innerHTML.includes('cancelBtnCls')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: async', async () => {
		const mockFn = vi.fn();
		let data = null;
		const instance = new KPopconfirmAsync({
			target: host
		});
		expect(instance).toBeTruthy();
		instance.$on('confirm', (e) => {
			mockFn();
			data = e.detail;
		});
		instance.$on('cancel', (e) => {
			mockFn();
			data = e.detail;
		});
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		const btnCancel = host.querySelector('.k-button--info__ghost');
		const btnConfirm = host.querySelector('.k-button--primary__fill');
		btnConfirm.click();
		btnCancel.click();
		await tick();
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		expect(mockFn).toBeCalledTimes(2);
		expect(data).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('events: change', async () => {
		const mockFn = vi.fn();
		let show = false;
		const instance = new KPopconfirmEvtChange({
			target: host
		});
		expect(instance).toBeTruthy();
		instance.$on('change', (e) => {
			mockFn();
			show = e.detail;
		});
		const elm = host.querySelector('#test_trigger');
		elm.dispatchEvent(new Event('click', { bubbles: true }));
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		expect(mockFn).toBeCalledTimes(1);
		expect(show).toBeTruthy();
		const btnConfirm = host.querySelector('.k-button--primary__fill');
		btnConfirm?.dispatchEvent(
			new Event('click', {
				bubbles: true
			})
		);
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(mockFn).toBeCalledTimes(2);
		expect(show).not.toBeTruthy();
	});

	test('events: cancel', async () => {
		const mockFn = vi.fn();
		let show = null;
		const instance = new KPopconfirmEvtCancel({
			target: host
		});
		expect(instance).toBeTruthy();
		instance.$on('cancel', (e) => {
			mockFn();
			show = e.detail;
		});
		const elm = host.querySelector('#test_trigger');
		elm.dispatchEvent(new Event('click', { bubbles: true }));
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		expect(mockFn).toBeCalledTimes(0);
		expect(show).not.toBeTruthy();
		const btnCancel = host.querySelector('.k-button--info__ghost');
		btnCancel?.dispatchEvent(
			new Event('click', {
				bubbles: true
			})
		);
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(mockFn).toBeCalledTimes(1);
		expect(show).toBeTruthy();
		expect((show! as { type: string }).type === 'cancel').toBeTruthy();
	});

	test('events: confirm', async () => {
		const mockFn = vi.fn();
		let show = null;
		const instance = new KPopconfirmEvtConfirm({
			target: host
		});
		expect(instance).toBeTruthy();
		instance.$on('confirm', (e) => {
			mockFn();
			show = e.detail;
		});
		const elm = host.querySelector('#test_trigger');
		elm.dispatchEvent(new Event('click', { bubbles: true }));
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(host.innerHTML.includes('有美一人')).toBeTruthy();
		expect(
			host.innerHTML.includes('<div class="k-popconfirm--content">清扬婉兮</div>')
		).toBeTruthy();
		expect(mockFn).toBeCalledTimes(0);
		expect(show).not.toBeTruthy();
		const btnConfirm = host.querySelector('.k-button--primary__fill');
		btnConfirm?.dispatchEvent(
			new Event('click', {
				bubbles: true
			})
		);
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(mockFn).toBeCalledTimes(1);
		expect(show).toBeTruthy();
		expect((show! as { type: string }).type === 'confirm').toBeTruthy();
	});

	test('slot: title', async () => {
		const instance = new KPopconfirmSlotTitle({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('test title')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slot: content', async () => {
		const instance = new KPopconfirmSlotContent({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('test content')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slot: footer', async () => {
		const instance = new KPopconfirmSlotFooter({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const elm = host.querySelector('#test_trigger');
		elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('test footer')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
