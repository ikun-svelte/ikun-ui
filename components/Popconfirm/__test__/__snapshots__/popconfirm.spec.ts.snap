// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KPopconfirm > props: async 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="top" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both; bottom: 8px;" data-popper-reference-hidden="" data-popper-escaped=""><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div class="k-popconfirm--title"><span class="k-icon--base k-icon--base__dark k-icon-transition k-popconfirm--icon--info k-popconfirm--icon k-popconfirm--info " role="" aria-hidden="true"></span> 有美一人</div> <div class="k-popconfirm--content">清扬婉兮</div> <div class="k-popconfirm--footer"><button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--info__ghost k-button--info__ghost--dark k-button--info__active k-button--info__focus k-button--sm k-button--info__border k-button--info__animate mr-2 " aria-hidden="true">  Cancel </button> <button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm k-button--primary__animate" aria-hidden="true">  Confirm </button></div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;

exports[`Test: KPopconfirm > props: base 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both;"><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div class="k-popconfirm--title"><span class="k-icon--base k-icon--base__dark k-icon-transition k-popconfirm--icon--info k-popconfirm--icon k-popconfirm--info " role="" aria-hidden="true"></span> 有美一人</div> <div class="k-popconfirm--content">清扬婉兮</div> <div class="k-popconfirm--footer"><button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--info__ghost k-button--info__ghost--dark k-button--info__active k-button--info__focus k-button--sm k-button--info__border mr-2 " aria-hidden="true">  Cancel </button> <button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm" aria-hidden="true">  Confirm </button></div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;

exports[`Test: KPopconfirm > props: confirmBtnText & confirmBtnCls & cancelBtnText & cancelBtnCls 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both;"><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div class="k-popconfirm--title"><span class="k-icon--base k-icon--base__dark k-icon-transition k-popconfirm--icon--info k-popconfirm--icon k-popconfirm--info " role="" aria-hidden="true"></span> 有美一人</div> <div class="k-popconfirm--content">清扬婉兮</div> <div class="k-popconfirm--footer"><button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--info__ghost k-button--info__ghost--dark k-button--info__active k-button--info__focus k-button--sm k-button--info__border mr-2 cancelBtnCls" aria-hidden="true">  cancelBtnText </button> <button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm confirmBtnCls" aria-hidden="true">  confirmBtnText </button></div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;

exports[`Test: KPopconfirm > props: disabled 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> "`;

exports[`Test: KPopconfirm > props: icon 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both;"><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div class="k-popconfirm--title"><span class="k-icon--base k-icon--base__dark k-icon-transition i-carbon-settings k-popconfirm--icon " role="" aria-hidden="true"></span> 有美一人</div> <div class="k-popconfirm--content">清扬婉兮</div> <div class="k-popconfirm--footer"><button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--info__ghost k-button--info__ghost--dark k-button--info__active k-button--info__focus k-button--sm k-button--info__border mr-2 " aria-hidden="true">  Cancel </button> <button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm" aria-hidden="true">  Confirm </button></div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;

exports[`Test: KPopconfirm > props: showCancel 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both;"><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div class="k-popconfirm--title"><span class="k-icon--base k-icon--base__dark k-icon-transition k-popconfirm--icon--info k-popconfirm--icon k-popconfirm--info " role="" aria-hidden="true"></span> 有美一人</div> <div class="k-popconfirm--content">清扬婉兮</div> <div class="k-popconfirm--footer"> <button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm" aria-hidden="true">  Confirm </button></div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;

exports[`Test: KPopconfirm > slot: content 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both;"><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div class="k-popconfirm--title"><span class="k-icon--base k-icon--base__dark k-icon-transition k-popconfirm--icon--info k-popconfirm--icon k-popconfirm--info " role="" aria-hidden="true"></span> 有美一人</div> <div slot="content">test content</div> <div class="k-popconfirm--footer"><button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--info__ghost k-button--info__ghost--dark k-button--info__active k-button--info__focus k-button--sm k-button--info__border mr-2 " aria-hidden="true">  Cancel </button> <button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm" aria-hidden="true">  Confirm </button></div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;

exports[`Test: KPopconfirm > slot: footer 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both;"><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div class="k-popconfirm--title"><span class="k-icon--base k-icon--base__dark k-icon-transition k-popconfirm--icon--info k-popconfirm--icon k-popconfirm--info " role="" aria-hidden="true"></span> 有美一人</div> <div class="k-popconfirm--content">清扬婉兮</div> <div slot="footer">test footer</div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;

exports[`Test: KPopconfirm > slot: title 1`] = `"<div aria-hidden="true" class="flex" data-popover-trigger=""><div id="test_trigger">trigger</div></div> <div class="k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-13arsd0" data-popper-placement="" aria-hidden="true" style="position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_3896445108_0 200ms linear 0ms 1 both;"><div class="k-popconfirm k-popconfirm--base" slot="contentEl"><div slot="title">test title</div> <div class="k-popconfirm--content">清扬婉兮</div> <div class="k-popconfirm--footer"><button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--info__ghost k-button--info__ghost--dark k-button--info__active k-button--info__focus k-button--sm k-button--info__border mr-2 " aria-hidden="true">  Cancel </button> <button style="border-radius: 4px; font-size: 12px;" class="k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm" aria-hidden="true">  Confirm </button></div></div> <div k-popover-arrow="true" data-popper-arrow-bottom="" data-popper-arrow-top="" data-popper-arrow-right="" data-popper-arrow-left="" class="svelte-13arsd0"></div></div>"`;
